README.md
# severo-http-connector

A easy HTTP<->Kafka connector with a complete config file and swagger generation.

![http kafka converter](https://raw.githubusercontent.com/lsevero/severo-http-connector/master/http-kafka.jpeg)

## Usage

Download the `severo-http-connector-<VERSION>-standalone.jar` from the releases tab and run the jar with the config files with the command:

```bash
java -Dlog4j.configurationFile=<path-to>/log4j2.xml -Dconfig=<path-to>/config.edn -jar severo-http-connector-<VERSION>-standalone.jar
```
Check the config files in the `example-config` folder.

The kafka producer will insert a `http-response-id` field to the root of the input payload to be sent to kafka on the `send-topic`.
The HTTP connection will keep waiting for a message on the `listen-topic` with the `http-response-id` field, when we receive the message it will be returned to http.
The consumer optionally accepts a `http-status` which is the status code to be passed on the http response, if no http-status is given will be returned a 200.

## `config.edn`

Explained config.edn:
```clojure
{;Default kafka configs
;The kafka section contain all the java properties to control both consumer and producer. The same as the property files.
;they can be overwritten per route on the 'routes' section
:kafka {:consumer {"bootstrap.servers" "localhost:9092" 
                    "group.id" "test"
                    "auto.offset.reset" "latest"
                    "enable.auto.commit" "true"
                    "key.deserializer" "org.apache.kafka.common.serialization.StringDeserializer"
                    "value.deserializer" "org.apache.kafka.common.serialization.StringDeserializer"
                    }
         :producer {"bootstrap.servers" "localhost:9092"
                    "key.serializer" "org.apache.kafka.common.serialization.StringSerializer"
                    "value.serializer" "org.apache.kafka.common.serialization.StringSerializer"
                    }}
 ;Here we control some option of the http server
 :http {:min-threads 16
        :max-threads 100
        :max-idle-time 1800000 
        :request-header-size 8192
        :port 3000
        }
 ;Swagger options
 :swagger {:enabled? true ;swagger can be turned off setting this config to false. Defaults to true when it is not available.
           :title "SeveroHTTPConnector"
           :description "Connecting http to kafka"
           :path "/"}
 ;Any option that is not given at the `routes` is replaced to one the these default values
 :defaults {:send-topic "poc"
            :listen-topic "poc"
            :poll-duration 200
            :timeout 5000
            :partitions 1
            :replication 3
            }
 ;The list of all routes
 ;Each route can have multiple http methods with individual configs
 ;all of the options here are optional and if not given will default to the values above.
 :routes {"/test" {:post {:send-topic "poc"
                         :listen-topic "poc"
                         ;time to return a timeout on http
                         :timeout 2000 ;in milliseconds
                         ;The partitions and replication options are used to create the topic if it does not exist. If exists it will do nothing.
                         :partitions 1
                         :replication 3
                         ;Poll interval on the consumer loop
                         :poll-duration 100 ;in milisseconds 
                         :summary "test" ;swagger summary of the route
                         ;Here we can list and validate all the info given to the route.
                         ;This option is optional, if not given will pass the input to kafka without any validation
                         :parameters {:body ;Here we are using malli specs.
                                            ;These specs will be documented on swagger
                                            [:and
                                             [:map
                                              [:x [:and int? [:> 6]]]
                                              [:y number?]]
                                             [:fn (fn [{:keys [x y]}] (> x y))]]
                                      :query [:map
                                              [:id string?]]}
                         }
                  :get {:send-topic "poc2"
                        :listen-topic "poc2"
                        :timeout 200
                        :poll-duration 100
                        :parameters {:query [:map [:number [:and int? [:> 6]]]]}
                        :summary "testing get"
                        }}
          "/other-test/:id" {:post {:send-topic "poc3"
                                    :listen-topic "poc3"
                                    :timeout 2000
                                    :poll-duration 100;milliseconds
                                    :summary "more test"
                                    ;We can define kafka configs per route as well
                                    ;these maps will be merged against the kafka configs above, per-route configs prevail
                                    :consumer {"group.id" "other-test"
                                              }
                                    :producer {}
                                 }}
          }
 }

```

## Logging

This project is using Log4j2 to log, and accepts a log4j2 compatible config file.
Besides the logs inside SeveroHTTPConnector you can extract the logs inside the kafka and http library, although they are very verbose.
Check the `example-config/log4j2.xml` file for a example.
The logging config is optional, although is recommended to use.

## Building

You'll need [leiningen](https://leiningen.org/) installed, then:
```bash
lein uberjar
```

## License

Copyright © 2021 Lucas Severo

This program and the accompanying materials are made available under the
terms of the Eclipse Public License 2.0 which is available at
http://www.eclipse.org/legal/epl-2.0.

This Source Code may also be made available under the following Secondary
Licenses when the conditions for such availability set forth in the Eclipse
Public License, v. 2.0 are satisfied: GNU General Public License as published by
the Free Software Foundation, either version 2 of the License, or (at your
option) any later version, with the GNU Classpath Exception which is available
at https://www.gnu.org/software/classpath/license.html.


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file. This change log follows the conventions of [keepachangelog.com](http://keepachangelog.com/).

## [Unreleased]
### Changed
- Add a new arity to `make-widget-async` to provide a different widget shape.

## [0.1.1] - 2021-05-05
### Changed
- Documentation on how to make the widgets.

### Removed
- `make-widget-sync` - we're all async, all the time.

### Fixed
- Fixed widget maker to keep working when daylight savings switches over.

## 0.1.0 - 2021-05-05
### Added
- Files from the new template.
- Widget maker public API - `make-widget-sync`.

[Unreleased]: https://github.com/your-name/severo-http-connector/compare/0.1.1...HEAD
[0.1.1]: https://github.com/your-name/severo-http-connector/compare/0.1.0...0.1.1


src/severo_http_connector/core.clj
(ns severo-http-connector.core
  (:require
    [cheshire.core :as json]
    [clojure.core.async
     :as a
     :refer [>! <! >!! <!! go chan buffer close! thread go-loop
             alts! alts!! timeout]]
    [clojure.tools.logging :as log] 
    [config.core :refer [env]]
    [reitit.ring :as ring]
    [reitit.http :as http]
    [reitit.coercion.spec]
    [reitit.swagger :as swagger]
    [reitit.swagger-ui :as swagger-ui]
    [reitit.http.coercion :as coercion]
    [reitit.dev.pretty :as pretty]
    [reitit.interceptor.sieppari :as sieppari]
    [reitit.http.interceptors.parameters :as parameters]
    [reitit.http.interceptors.muuntaja :as muuntaja]
    [reitit.http.interceptors.exception :as exception]
    [reitit.http.interceptors.multipart :as multipart]
    [reitit.coercion.malli]
    [reitit.ring.malli]
    [malli.util :as mu]
    [malli.core :as malli]
    [malli.error :as me]
    [muuntaja.core :as m]
    [ring.adapter.jetty :as jetty]
    [severo-http-connector
     [consumer :refer [consumer!]]
     [producer :refer [producer!]]])
  (:import
    [java.util UUID])
  (:gen-class))

(def get-route-spec
  [:map {:closed true}
   [:send-topic {:optional true} string?]
   [:listen-topic {:optional true} string?]
   [:poll-duration {:optional true} pos-int?]
   [:timeout {:optional true} pos-int?]
   [:partitions {:optional true} pos-int?]
   [:replication {:optional true} pos-int?]
   [:summary {:optional true} string?]
   [:consumer {:optional true} [:map-of :string :string]]
   [:producer {:optional true} [:map-of :string :string]]
   [:parameters {:optional true} [:map {:closed true}
                                  [:query any?]]]])

(def post-route-spec
  [:map {:closed true}
   [:send-topic {:optional true} string?]
   [:listen-topic {:optional true} string?]
   [:poll-duration {:optional true} pos-int?]
   [:timeout {:optional true} pos-int?]
   [:partitions {:optional true} pos-int?]
   [:replication {:optional true} pos-int?]
   [:summary {:optional true} string?]
   [:consumer {:optional true} [:map-of :string :string]]
   [:producer {:optional true} [:map-of :string :string]]
   [:parameters {:optional true} [:map {:closed true}
                                  [:query any?]
                                  [:body any?]]]])

(def config-spec
  [:map
   [:kafka [:map {:closed true}
            [:consumer [:map-of :string :string]]
            [:producer [:map-of :string :string]]
            ]]
   [:http [:and
           [:map
            [:min-threads {:optional true} pos-int?]
            [:max-threads {:optional true} pos-int?]
            [:max-idle-time {:optional true} pos-int?]
            [:request-header-size {:optional true} pos-int?]
            [:port {:optional true} [:and pos-int? [:< 65535]]]
            ]
           [:fn {:error/message "max-threads need to be greater than min-threads"} (fn [{:keys [min-threads max-threads]}] (> max-threads min-threads))]
           ]]
   [:swagger [:map {:closed true}
              [:enabled? {:optional true} boolean?]
              [:title {:optional true} string?]
              [:description {:optional true} string?]
              [:path {:optional true} string?]
              ]]
   [:defaults [:map {:closed true}
               [:send-topic {:optional true} string?]
               [:listen-topic {:optional true} string?]
               [:poll-duration {:optional true} pos-int?]
               [:timeout {:optional true} pos-int?]
               [:partitions {:optional true} pos-int?]
               [:replication {:optional true} pos-int?]
               ]]
   [:routes 
    [:and [:map-of :string [:and
                            [:map {:closed true}
                             [:get {:optional true} get-route-spec]
                             [:head {:optional true} get-route-spec]
                             [:post {:optional true} post-route-spec]
                             [:put {:optional true} post-route-spec]
                             [:delete {:optional true} post-route-spec]
                             [:connect {:optional true} post-route-spec]
                             [:options {:optional true} post-route-spec]
                             [:trace {:optional true} post-route-spec]
                             [:patch {:optional true} post-route-spec]]
                            [:fn {:error/message "method map cannot be empty"} (fn [m] (not (empty? m)))]]]
     [:fn {:error/message "route map cannot be empty"} (fn [m] (not (empty? m)))]]]])

(defn create-generic-handler
  [send-topic listen-topic timeout-ms poll-duration partitions replication consumer-cfg producer-cfg]
  (let [canal-producer (chan)
        cache (atom {})]
    (consumer! listen-topic consumer-cfg cache :duration poll-duration)
    (producer! send-topic partitions replication producer-cfg canal-producer)
    (fn [{:keys [parameters headers] :as req}]
      (log/trace "http-request: " req)
      (let [uuid (str (UUID/randomUUID))
            payload (-> parameters
                      (assoc :http-response-id uuid)
                      (assoc :headers headers)
                      (assoc :uri (:uri req)))
            canal-resposta (chan)]
        (swap! cache assoc uuid canal-resposta)
        (log/trace "cache: " @cache)
        (>!! canal-producer (json/generate-string payload))
        (let [[value channel] (alts!! [canal-resposta (timeout timeout-ms)])
              _ (swap! cache dissoc uuid)]
          (if value 
            (let [{:keys [http-status] :or {http-status 200}} value]
              (log/trace "consumed payload: " value)
              {:status http-status
               :headers {"Content-Type" "application/json"}
               :body (json/generate-string (apply dissoc value [:http-status :http-response-id]))})
            {:status 504
             :headers {"Content-Type" "application/json"}
             :body (json/generate-string {:message "Timeout."})}))))))

(defn prepare-reitit-handlers
  []
  (into [""
         {:swagger {:tags ["api"]}}]
        (mapv (fn [[route method-map]]
                [route (into {}
                             (mapv (fn [[method {:keys [send-topic listen-topic timeout poll-duration
                                                        partitions replication consumer producer
                                                        parameters] :as conf}]]
                                     (let [conf-aux (-> conf
                                                        (dissoc :send-topic :listen-topic :timeout :poll-duration :partitions :replication :consumer :producer)
                                                        (assoc :handler (create-generic-handler (or send-topic
                                                                                                    (-> env :defaults :send-topic)
                                                                                                    (throw (ex-info "send-topic cannot be null" {})))
                                                                                                (or listen-topic
                                                                                                    (-> env :defaults :listen-topic)
                                                                                                    (throw (ex-info "listen-topic cannot be null" {})))
                                                                                                (or timeout
                                                                                                    (-> env :defaults :timeout)
                                                                                                    (throw (ex-info "timeout cannot be null" {})))
                                                                                                (or poll-duration
                                                                                                    (-> env :defaults :poll-duration)
                                                                                                    (throw (ex-info "poll-duration cannot be null" {})))
                                                                                                (or partitions
                                                                                                    (-> env :defaults :partitions)
                                                                                                    (throw (ex-info "partitions cannot be null" {})))
                                                                                                (or replication 
                                                                                                    (-> env :defaults :replication)
                                                                                                    (throw (ex-info "replication" {})))
                                                                                                (or consumer {})
                                                                                                (or producer {}))))
                                           conf-final (if parameters
                                                        conf-aux
                                                        (assoc conf-aux :parameters (if (#{:get :head} method)
                                                                                      {:query map?
                                                                                       :path map?
                                                                                       }
                                                                                      {:query map?
                                                                                       :path map?
                                                                                       :body map?})))]

                                       [method conf-final]))
                                   method-map))])
              (:routes env))))

(defn swagger-http-server 
  []
  (let [app (http/ring-handler
              (http/router
                [["/swagger.json"
                  {:get {:no-doc true
                         :swagger {:info {:title (or (-> env :swagger :title) "SeveroHTTPConnector")
                                          :description (or (-> env :swagger :description) "Connecting http to kafka")}}
                         :handler (swagger/create-swagger-handler)}}]
                 (prepare-reitit-handlers)
                 ]

                {;:reitit.interceptor/transform dev/print-context-diffs ;; pretty context diffs
                 ;;:validate spec/validate ;; enable spec validation for route data
                 ;;:reitit.spec/wrap spell/closed ;; strict top-level validation
                 :exception pretty/exception
                 :data {:coercion (reitit.coercion.malli/create
                                    {;; set of keys to include in error messages
                                     :error-keys #{#_:type :coercion :in :schema :value :errors :humanized #_:transformed}
                                     ;; schema identity function (default: close all map schemas)
                                     :compile mu/closed-schema
                                     ;; strip-extra-keys (effects only predefined transformers)
                                     :strip-extra-keys true
                                     ;; add/set default values
                                     :default-values true
                                     ;; malli options
                                     :options nil})
                        :muuntaja m/instance
                        :interceptors [;; swagger feature
                                       swagger/swagger-feature
                                       ;; query-params & form-params
                                       (parameters/parameters-interceptor)
                                       ;; content-negotiation
                                       (muuntaja/format-negotiate-interceptor)
                                       ;; encoding response body
                                       (muuntaja/format-response-interceptor)
                                       ;; exception handling
                                       (exception/exception-interceptor)
                                       ;; decoding request body
                                       (muuntaja/format-request-interceptor)
                                       ;; coercing response bodys
                                       (coercion/coerce-response-interceptor)
                                       ;; coercing request parameters
                                       (coercion/coerce-request-interceptor)
                                       ;; multipart
                                       (multipart/multipart-interceptor)]}})
              (ring/routes
                (swagger-ui/create-swagger-ui-handler
                  {:path (or (-> env :swagger :path) "/")
                   :config {:validatorUrl nil
                            :operationsSorter "alpha"}})
                (ring/create-default-handler
                  {:not-found (constantly {:status 404
                                           :headers {"Content-Type" "application/json"}
                                           :body (json/generate-string {:message "Not found"})})
                   :method-not-allowed (constantly {:status 405
                                                    :headers {"Content-Type" "application/json"}
                                                    :body (json/generate-string {:message "Method not allowed"})})
                   :not-acceptable (constantly {:status 406
                                                :headers {"Content-Type" "application/json"}
                                                :body (json/generate-string {:message "Not acceptable"})})
                   }))
              {:executor sieppari/executor})]
    (jetty/run-jetty app {:port (or (-> env :http :port) 3000)
                          :join? true
                          :min-threads (or (-> env :http :min-threads) 8)
                          :max-threads (or (-> env :http :max-threads) 20)
                          :max-idle-time (or (-> env :http :max-idle-time) (* 60 1000 10))
                          :request-header-size (or (-> env :http :request-header-size) 8192)
                          })))

(defn http-server 
  []
  (let [app (http/ring-handler
              (http/router
                [(prepare-reitit-handlers)]

                {;:reitit.interceptor/transform dev/print-context-diffs ;; pretty context diffs
                 ;;:validate spec/validate ;; enable spec validation for route data
                 ;;:reitit.spec/wrap spell/closed ;; strict top-level validation
                 :exception pretty/exception
                 :data {:coercion (reitit.coercion.malli/create
                                    {;; set of keys to include in error messages
                                     :error-keys #{#_:type :coercion :in :schema :value :errors :humanized #_:transformed}
                                     ;; schema identity function (default: close all map schemas)
                                     :compile mu/closed-schema
                                     ;; strip-extra-keys (effects only predefined transformers)
                                     :strip-extra-keys true
                                     ;; add/set default values
                                     :default-values true
                                     ;; malli options
                                     :options nil})
                        :muuntaja m/instance
                        :interceptors [;; swagger feature
                                       swagger/swagger-feature
                                       ;; query-params & form-params
                                       (parameters/parameters-interceptor)
                                       ;; content-negotiation
                                       (muuntaja/format-negotiate-interceptor)
                                       ;; encoding response body
                                       (muuntaja/format-response-interceptor)
                                       ;; exception handling
                                       (exception/exception-interceptor)
                                       ;; decoding request body
                                       (muuntaja/format-request-interceptor)
                                       ;; coercing response bodys
                                       (coercion/coerce-response-interceptor)
                                       ;; coercing request parameters
                                       (coercion/coerce-request-interceptor)
                                       ;; multipart
                                       (multipart/multipart-interceptor)]}})
              (ring/routes
                (ring/create-default-handler
                  {:not-found (constantly {:status 404
                                           :headers {"Content-Type" "application/json"}
                                           :body (json/generate-string {:message "Not found"})})
                   :method-not-allowed (constantly {:status 405
                                                    :headers {"Content-Type" "application/json"}
                                                    :body (json/generate-string {:message "Method not allowed"})})
                   :not-acceptable (constantly {:status 406
                                                :headers {"Content-Type" "application/json"}
                                                :body (json/generate-string {:message "Not acceptable"})})
                   }))
              {:executor sieppari/executor})]
    (jetty/run-jetty app {:port (or (-> env :http :port) 3000)
                          :join? true
                          :min-threads (or (-> env :http :min-threads) 8)
                          :max-threads (or (-> env :http :max-threads) 20)
                          :max-idle-time (or (-> env :http :max-idle-time) (* 60 1000 10))
                          :request-header-size (or (-> env :http :request-header-size) 8192)
                          })))

(defn -main [& args]
  (try
    (log/info "Reading the config file...")
    (when-not (malli/validate config-spec env)
      (log/error (apply str (repeat 120 "=")))
      (log/error "Config file not valid!!!!")
      (log/error (-> config-spec
                   (malli/explain env)
                   me/with-spell-checking
                   me/humanize))
      (log/error (apply str (repeat 120 "=")))
      (System/exit 1))
    (catch Exception e
      (do (log/error e "Error validating the config file")
          (System/exit 1))))
  (try
    (log/info "Config file is valid")
    (log/info "Initing web server")
    (if (false? (-> env :swagger :enabled?))
      (http-server)
      (swagger-http-server))
    (catch Exception e
      (do (log/error e "Error during the http server init.")
          (System/exit 1))))) 
